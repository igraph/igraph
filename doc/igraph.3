.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.\" Copyright (C) 2006-2021  The igraph development team
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH IGRAPH 3 "May 2021" "igraph library"
.SH NAME
igraph \- a library for creating and manipulating graphs
.SH DESCRIPTION
.B igraph
is a C library for complex network analysis and graph theory, with emphasis on
efficiency, portability and ease of use.
.SH DOCUMENTATION
The full documentation can be downloaded from the homepage of the
library:
.RI < https://igraph.org/c/doc >
.SH BUGS
If you think you have found a bug in igraph, feel free to file a bug report
in the issue tracker at:
.RI < https://github.com/igraph/igraph/issues >

.SH AUTHORS
Gabor Csardi <csardi.gabor@gmail.com>,
.br
Tamas Nepusz <ntamas@gmail.com>,
.br
Vincent Traag <vtraag@gmail.com>,
.br
Szabolcs Horvat <szhorvat@gmail.com>,
.br
Fabio Zanini <fabio.zanini@fastmail.fm>,
.br
Daniel Noom <ggatw@outlook.com>
