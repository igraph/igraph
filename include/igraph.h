/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2003-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_H
#define IGRAPH_H

#ifndef _GNU_SOURCE
    #define _GNU_SOURCE 1
#endif

#include "igraph_version.h"
#include "igraph_memory.h"
#include "igraph_error.h"
#include "igraph_random.h"
#include "igraph_progress.h"
#include "igraph_statusbar.h"

#include "igraph_types.h"
#include "igraph_complex.h"
#include "igraph_vector.h"
#include "igraph_matrix.h"
#include "igraph_array.h"
#include "igraph_dqueue.h"
#include "igraph_stack.h"
#include "igraph_heap.h"
#include "igraph_psumtree.h"
#include "igraph_strvector.h"
#include "igraph_vector_ptr.h"
#include "igraph_spmatrix.h"
#include "igraph_sparsemat.h"
#include "igraph_qsort.h"

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_iterators.h"
#include "igraph_interface.h"
#include "igraph_constructors.h"
#include "igraph_games.h"
#include "igraph_microscopic_update.h"
#include "igraph_centrality.h"
#include "igraph_paths.h"
#include "igraph_components.h"
#include "igraph_structural.h"
#include "igraph_transitivity.h"
#include "igraph_neighborhood.h"
#include "igraph_topology.h"
#include "igraph_bipartite.h"
#include "igraph_cliques.h"
#include "igraph_layout.h"
#include "igraph_visitor.h"
#include "igraph_community.h"
#include "igraph_conversion.h"
#include "igraph_foreign.h"
#include "igraph_motifs.h"
#include "igraph_operators.h"
#include "igraph_flow.h"
#include "igraph_nongraph.h"
#include "igraph_cocitation.h"
#include "igraph_adjlist.h"
#include "igraph_attributes.h"
#include "igraph_blas.h"
#include "igraph_lapack.h"
#include "igraph_arpack.h"
#include "igraph_mixing.h"
#include "igraph_separators.h"
#include "igraph_cohesive_blocks.h"
#include "igraph_eigen.h"
#include "igraph_hrg.h"
#include "igraph_threading.h"
#include "igraph_interrupt.h"
#include "igraph_scg.h"
#include "igraph_matching.h"
#include "igraph_embedding.h"
#include "igraph_scan.h"
#include "igraph_graphlets.h"
#include "igraph_epidemics.h"
#include "igraph_lsap.h"
#include "igraph_coloring.h"

#endif
