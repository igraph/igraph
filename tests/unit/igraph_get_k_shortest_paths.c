/*
   IGraph library.
   Copyright (C) 2021  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>
#include "test_utilities.h"

void call_and_print(
    igraph_t *graph, igraph_vector_t *weights, igraph_integer_t k,
    igraph_integer_t from, igraph_integer_t to, igraph_neimode_t mode
) {
    igraph_vector_int_list_t vertex_paths;
    igraph_vector_int_list_t edge_paths;
    igraph_vector_int_list_t paths_for_verification;
    igraph_integer_t i, n;

    igraph_vector_int_list_init(&vertex_paths, 0);
    igraph_vector_int_list_init(&edge_paths, 0);
    igraph_vector_int_list_init(&paths_for_verification, 0);

    IGRAPH_ASSERT(igraph_get_k_shortest_paths(graph, weights, &vertex_paths, &edge_paths, k, from, to, mode) == IGRAPH_SUCCESS);

    printf("result (vertex IDs): \n");
    print_vector_int_list(&vertex_paths);

    printf("result (edge IDs): \n");
    print_vector_int_list(&edge_paths);

    /* now we execute the same test but with only one of vertex_paths or edge_paths,
     * and we check that we get the same result */
    IGRAPH_ASSERT(igraph_get_k_shortest_paths(graph, weights, NULL, &paths_for_verification, k, from, to, mode) == IGRAPH_SUCCESS);
    n = igraph_vector_int_list_size(&paths_for_verification);
    IGRAPH_ASSERT(n == igraph_vector_int_list_size(&edge_paths));
    for (i = 0; i < n; i++) {
        IGRAPH_ASSERT(igraph_vector_int_is_equal(
            igraph_vector_int_list_get_ptr(&edge_paths, i),
            igraph_vector_int_list_get_ptr(&paths_for_verification, i)
        ));
    }

    IGRAPH_ASSERT(igraph_get_k_shortest_paths(graph, weights, &paths_for_verification, NULL, k, from, to, mode) == IGRAPH_SUCCESS);
    n = igraph_vector_int_list_size(&paths_for_verification);
    IGRAPH_ASSERT(n == igraph_vector_int_list_size(&vertex_paths));
    for (i = 0; i < n; i++) {
        IGRAPH_ASSERT(igraph_vector_int_is_equal(
            igraph_vector_int_list_get_ptr(&vertex_paths, i),
            igraph_vector_int_list_get_ptr(&paths_for_verification, i)
        ));
    }

    igraph_vector_int_list_destroy(&paths_for_verification);
    igraph_vector_int_list_destroy(&vertex_paths);
    igraph_vector_int_list_destroy(&edge_paths);
    printf("\n");

    VERIFY_FINALLY_STACK();
}


int main(void) {
    /* Wiki example taken from https://en.wikipedia.org/wiki/Yen's_algorithm */
    igraph_t g_0, g_1, g_2, g_2c, g_wiki, g_wiki_u, g_sz, g_lattice;
    igraph_vector_t weights, weights_wiki, weights_inf;
    igraph_vector_int_t dims;
    igraph_vector_int_list_t paths;

    igraph_vector_int_list_init(&paths, 0);

    igraph_small(&g_0, 0, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_1, 1, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_2, 2, IGRAPH_UNDIRECTED, -1);
    igraph_small(&g_2c, 2, IGRAPH_UNDIRECTED, 0,1, -1);
    igraph_small(&g_wiki, 6, IGRAPH_DIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
    igraph_small(&g_wiki_u, 6, IGRAPH_UNDIRECTED, 0,1, 0,2, 1,3, 2,1, 2,3, 2,4, 3,4, 3,5, 4,5, -1);
    igraph_small(&g_sz, 5, IGRAPH_DIRECTED, 0, 1, 1, 2, 1, 2, 2, 0, 0, 3, 3, 4, 4, 2, 2, 2, 2, 4, 0, 4, -1);

    igraph_vector_int_init_int(&dims, 2, 3, 3);
    igraph_square_lattice(&g_lattice, &dims, 1, IGRAPH_UNDIRECTED, /* mutual = */ 0, /* periodic = */ NULL);
    igraph_vector_int_destroy(&dims);

    igraph_vector_init(&weights, 0);
    igraph_vector_init_int(&weights_wiki, 9, 3, 2, 4, 1, 2, 3, 2, 1, 2);
    igraph_vector_init_real(&weights_inf, 1, IGRAPH_INFINITY);

    printf("One vertex:\n");
    call_and_print(&g_1, &weights, 2, 0, 0, IGRAPH_ALL);

    printf("Two disconnected vertices:\n");
    call_and_print(&g_2, &weights, 2, 0, 1, IGRAPH_ALL);

    printf("Two vertices with infinite weight edge in between:\n");
    call_and_print(&g_2c, &weights_inf, 2, 0, 1, IGRAPH_ALL);

    printf("Wiki example:\n");
    call_and_print(&g_wiki, &weights_wiki, 10, 0, 5, IGRAPH_OUT);

    printf("Wiki example, 0 shortest paths:\n");
    call_and_print(&g_wiki, &weights_wiki, 0, 0, 5, IGRAPH_OUT);

    printf("Wiki example, 2 shortest paths:\n");
    call_and_print(&g_wiki, &weights_wiki, 2, 0, 5, IGRAPH_OUT);

    printf("Wiki example, other direction:\n");
    call_and_print(&g_wiki, &weights_wiki, 10, 5, 0, IGRAPH_IN);

    printf("Wiki example, direction ignored:\n");
    call_and_print(&g_wiki, &weights_wiki, 20, 5, 0, IGRAPH_ALL);

    printf("Wiki example, undirected:\n");
    call_and_print(&g_wiki_u, &weights_wiki, 20, 5, 0, IGRAPH_ALL);

    printf("Wiki example, no weights:\n");
    call_and_print(&g_wiki, NULL, 10, 0, 5, IGRAPH_OUT);

    printf("Directed unweighted graph:\n");
    call_and_print(&g_sz, NULL, 4, 0, 4, IGRAPH_OUT);

    printf("3x3 square lattice:\n");
    call_and_print(&g_lattice, NULL, 6, 0, 8, IGRAPH_OUT);

    printf("Zero vertices, from and to don't exist:\n");
    CHECK_ERROR(igraph_get_k_shortest_paths(&g_0, &weights, NULL, &paths, 4, 0, 0, IGRAPH_ALL), IGRAPH_EINVVID);

    printf("Wrong weights length:\n");
    CHECK_ERROR(igraph_get_k_shortest_paths(&g_wiki, &weights, NULL, &paths, 4, 0, 5, IGRAPH_ALL), IGRAPH_EINVAL);

    printf("Non-existent mode:\n");
    CHECK_ERROR(igraph_get_k_shortest_paths(&g_1, &weights, NULL, &paths, 4, 0, 0, (igraph_neimode_t) 100), IGRAPH_EINVMODE);

    igraph_destroy(&g_0);
    igraph_destroy(&g_1);
    igraph_destroy(&g_2);
    igraph_destroy(&g_2c);
    igraph_destroy(&g_wiki);
    igraph_destroy(&g_wiki_u);
    igraph_destroy(&g_sz);
    igraph_destroy(&g_lattice);
    igraph_vector_destroy(&weights);
    igraph_vector_destroy(&weights_wiki);
    igraph_vector_destroy(&weights_inf);
    igraph_vector_int_list_destroy(&paths);

    VERIFY_FINALLY_STACK();
    return 0;
}
