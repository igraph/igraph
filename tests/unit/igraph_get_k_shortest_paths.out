One vertex:
result (vertex IDs): 
{
  0: ( 0 )
}
result (edge IDs): 
{
  0: ( )
}

Two disconnected vertices:
result (vertex IDs): 
{
}
result (edge IDs): 
{
}

Two vertices with infinite weight edge in between:
result (vertex IDs): 
{
}
result (edge IDs): 
{
}

Wiki example:
result (vertex IDs): 
{
  0: ( 0 2 3 5 )
  1: ( 0 2 4 5 )
  2: ( 0 1 3 5 )
  3: ( 0 2 1 3 5 )
  4: ( 0 2 3 4 5 )
  5: ( 0 1 3 4 5 )
  6: ( 0 2 1 3 4 5 )
}
result (edge IDs): 
{
  0: ( 1 4 7 )
  1: ( 1 5 8 )
  2: ( 0 2 7 )
  3: ( 1 3 2 7 )
  4: ( 1 4 6 8 )
  5: ( 0 2 6 8 )
  6: ( 1 3 2 6 8 )
}

Wiki example, 0 shortest paths:
result (vertex IDs): 
{
}
result (edge IDs): 
{
}

Wiki example, 2 shortest paths:
result (vertex IDs): 
{
  0: ( 0 2 3 5 )
  1: ( 0 2 4 5 )
}
result (edge IDs): 
{
  0: ( 1 4 7 )
  1: ( 1 5 8 )
}

Wiki example, other direction:
result (vertex IDs): 
{
  0: ( 5 3 2 0 )
  1: ( 5 4 2 0 )
  2: ( 5 3 1 0 )
  3: ( 5 4 3 2 0 )
  4: ( 5 3 1 2 0 )
  5: ( 5 4 3 1 0 )
  6: ( 5 4 3 1 2 0 )
}
result (edge IDs): 
{
  0: ( 7 4 1 )
  1: ( 8 5 1 )
  2: ( 7 2 0 )
  3: ( 8 6 4 1 )
  4: ( 7 2 3 1 )
  5: ( 8 6 2 0 )
  6: ( 8 6 2 3 1 )
}

Wiki example, direction ignored:
result (vertex IDs): 
{
  0: ( 5 3 2 0 )
  1: ( 5 4 2 0 )
  2: ( 5 3 2 1 0 )
  3: ( 5 3 1 0 )
  4: ( 5 4 3 2 0 )
  5: ( 5 3 1 2 0 )
  6: ( 5 3 4 2 0 )
  7: ( 5 4 2 1 0 )
  8: ( 5 3 4 2 1 0 )
  9: ( 5 4 3 2 1 0 )
  10: ( 5 4 3 1 0 )
  11: ( 5 4 3 1 2 0 )
  12: ( 5 4 2 3 1 0 )
}
result (edge IDs): 
{
  0: ( 7 4 1 )
  1: ( 8 5 1 )
  2: ( 7 4 3 0 )
  3: ( 7 2 0 )
  4: ( 8 6 4 1 )
  5: ( 7 2 3 1 )
  6: ( 7 6 5 1 )
  7: ( 8 5 3 0 )
  8: ( 7 6 5 3 0 )
  9: ( 8 6 4 3 0 )
  10: ( 8 6 2 0 )
  11: ( 8 6 2 3 1 )
  12: ( 8 5 4 2 0 )
}

Wiki example, undirected:
result (vertex IDs): 
{
  0: ( 5 3 2 0 )
  1: ( 5 4 2 0 )
  2: ( 5 3 2 1 0 )
  3: ( 5 3 1 0 )
  4: ( 5 4 3 2 0 )
  5: ( 5 3 1 2 0 )
  6: ( 5 3 4 2 0 )
  7: ( 5 4 2 1 0 )
  8: ( 5 3 4 2 1 0 )
  9: ( 5 4 3 2 1 0 )
  10: ( 5 4 3 1 0 )
  11: ( 5 4 3 1 2 0 )
  12: ( 5 4 2 3 1 0 )
}
result (edge IDs): 
{
  0: ( 7 4 1 )
  1: ( 8 5 1 )
  2: ( 7 4 3 0 )
  3: ( 7 2 0 )
  4: ( 8 6 4 1 )
  5: ( 7 2 3 1 )
  6: ( 7 6 5 1 )
  7: ( 8 5 3 0 )
  8: ( 7 6 5 3 0 )
  9: ( 8 6 4 3 0 )
  10: ( 8 6 2 0 )
  11: ( 8 6 2 3 1 )
  12: ( 8 5 4 2 0 )
}

Wiki example, no weights:
result (vertex IDs): 
{
  0: ( 0 1 3 5 )
  1: ( 0 2 4 5 )
  2: ( 0 2 3 5 )
  3: ( 0 1 3 4 5 )
  4: ( 0 2 3 4 5 )
  5: ( 0 2 1 3 5 )
  6: ( 0 2 1 3 4 5 )
}
result (edge IDs): 
{
  0: ( 0 2 7 )
  1: ( 1 5 8 )
  2: ( 1 4 7 )
  3: ( 0 2 6 8 )
  4: ( 1 4 6 8 )
  5: ( 1 3 2 7 )
  6: ( 1 3 2 6 8 )
}

Directed unweighted graph:
result (vertex IDs): 
{
  0: ( 0 4 )
  1: ( 0 3 4 )
  2: ( 0 1 2 4 )
  3: ( 0 1 2 4 )
}
result (edge IDs): 
{
  0: ( 9 )
  1: ( 4 5 )
  2: ( 0 2 8 )
  3: ( 0 1 8 )
}

3x3 square lattice:
result (vertex IDs): 
{
  0: ( 0 1 2 5 8 )
  1: ( 0 3 4 5 8 )
  2: ( 0 1 4 7 8 )
  3: ( 0 3 4 7 8 )
  4: ( 0 1 4 5 8 )
  5: ( 0 3 6 7 8 )
}
result (edge IDs): 
{
  0: ( 0 2 4 9 )
  1: ( 1 5 7 9 )
  2: ( 0 3 8 11 )
  3: ( 1 5 8 11 )
  4: ( 0 3 7 9 )
  5: ( 1 6 10 11 )
}

Zero vertices, from and to don't exist:
Wrong weights length:
Non-existent mode:
